contracts/extensions/ExtensionManager.sol

```diff
@@ -0,0 +1,167 @@
+    require(resolvers[_extensionId][_version] == address(0x0), "extension-manager-already-added");
+    require(_resolver != address(0x0), "extension-manager-bad-resolver");
+
+    resolvers[_extensionId][_version] = _resolver;
+    roles[_extensionId][_version] = _roles;
+
+    emit ExtensionAdded(_extensionId, _version);
+  }
+
+  function installExtension(
+    bytes32 _extensionId,
+    uint256 _version,
+    address _colony,
+    uint256 _rootChildSkillIndex,
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _domainId
+  )
+    public
+    isAuthorized(_colony, _permissionDomainId, _childSkillIndex, _domainId)
+  {
+    require(
+      installations[_extensionId][_version][_colony][_domainId] == address(0x0),
+      "extension-manager-already-installed"
+    );
+
+    ColonyExtension extension = new ColonyExtension();
+    installations[_extensionId][_version][_colony][_domainId] = address(extension);
+
+    assignRoles(_colony, _rootChildSkillIndex, _domainId, _extensionId, _version, true);
+    address resolver = resolvers[_extensionId][_version];
+    extension.install(_colony, resolver);
+
+    emit ExtensionInstalled(_extensionId, _version, _colony, _domainId);
+  }
+
+  function uninstallExtension(
+    bytes32 _extensionId,
+    uint256 _version,
+    address payable _colony,
+    uint256 _rootChildSkillIndex,
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _domainId
+  )
+    public
+    isAuthorized(_colony, _permissionDomainId, _childSkillIndex, _domainId)
+  {
+    require(
+      installations[_extensionId][_version][_colony][_domainId] != address(0x0),
+      "extension-manager-not-installed"
+    );
+
+    ColonyExtension extension = ColonyExtension(installations[_extensionId][_version][_colony][_domainId]);
+    installations[_extensionId][_version][_colony][_domainId] = address(0x0);
+
+    assignRoles(_colony, _rootChildSkillIndex, _domainId, _extensionId, _version, false);
+    extension.uninstall(_colony);
+
+    emit ExtensionUninstalled(_extensionId, _version, _colony, _domainId);
+  }
+
+  function getResolver(bytes32 _extensionId, uint256 _version)
+    public
+    view
+    returns (address)
+  {
+    return resolvers[_extensionId][_version];
+  }
+
+  function getExtension(bytes32 _extensionId, uint256 _version, address _colony, uint256 _domainId)
+    public
+    view
+    returns (address)
+  {
+    return installations[_extensionId][_version][_colony][_domainId];
+  }
+
+  function assignRoles(
+    address _colony,
+    uint256 _rootChildSkillIndex,
+    uint256 _domainId,
+    bytes32 _extensionId,
+    uint256 _version,
+    bool _setTo
+  )
+    internal
+  {
+    IColony colony = IColony(_colony);
+    address extension = installations[_extensionId][_version][_colony][_domainId];
+    uint8[] storage extensionRoles = roles[_extensionId][_version];
+
+    for (uint256 i; i < extensionRoles.length; i++) {
```

Comment:
I do not like any of this loop one bit. I'm not sure how to make it better though.

Reply:
My reasoning was that since the only one who can input the array is the metaColony, there's very little risk.